subprojects/testing-jvm/src/main/java/org/gradle/api/internal/tasks/testing/junit/JUnitTestEventAdapter.java

```diff
@@ -29,20 +29,27 @@
 import org.junit.runner.notification.RunListener;
 
 import java.util.HashMap;
-import java.util.HashSet;
 import java.util.Map;
-import java.util.Set;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 
 public class JUnitTestEventAdapter extends RunListener {
+    private class JUnitTestEventInternalDescriptor {
+        public TestDescriptorInternal descriptor;
+        public TestResult.ResultType result;
+
+        public JUnitTestEventInternalDescriptor(TestDescriptorInternal descriptor, TestResult.ResultType result) {
+            this.descriptor = descriptor;
+            this.result = result;
+        }
+    }
+
     private static final Pattern DESCRIPTOR_PATTERN = Pattern.compile("(.*)\\((.*)\\)", Pattern.DOTALL);
     private final TestResultProcessor resultProcessor;
     private final Clock clock;
     private final IdGenerator<?> idGenerator;
     private final Object lock = new Object();
-    private final Map<Description, TestDescriptorInternal> executing = new HashMap<Description, TestDescriptorInternal>();
-    private final Set<Description> assumptionFailed = new HashSet<Description>();
+    private final Map<Integer, JUnitTestEventInternalDescriptor> executing = new HashMap<Integer, JUnitTestEventInternalDescriptor>();
```

Comment:
This is very wrong. `hashCode()` can't be used as a key because different `description`s might have the same hashCode. According to `Object.hashCode`'s javadoc:

> It is **not** required that if two objects are unequal
> according to the {@link java.lang.Object#equals(java.lang.Object)}
> method, then calling the {@code hashCode} method on each of the
> two objects must produce distinct integer results.

Reply:
`Description`'s `hashCode` implementation uses the `fUniqueId` field, the same `fUniqueId` field is used inside `Description`'s `equals` implementation. However, it is probably best to let `HashMap` call `Description`'s `equals` implementation like it did in the past, so I will change that.

Reply:
But `Description` is not a final class... how can you be so sure its `hashCode` won't be overwritten?